LCM/scripts/InstallModule.py

```diff
@@ -16,188 +25,293 @@ def usage():
+
+    # Open the zip file
+    moduleZipFile = zipfile.ZipFile(moduleZipFilePath)
+
+    try:
+        # Check that the zip file members all have valid CRC's and file headers
+        firstBadFileName = moduleZipFile.testzip()
+        if firstBadFileName is not None:
+            exitWithError("The provided module zip file contains at least one invalid file. The first invalid file is named " + firstBadFileName + ". Please ensure that the file has a valid CRC and file header.")
+
+        # Check if the module was zipped to include the module folder or just the module folder's contents.
+        zipIncludesModuleFolder = False
+
+        for moduleZipMemberName in moduleZipFile.namelist():
+            if moduleZipMemberName.startswith('/') or moduleZipMemberName.startswith('..'):
+                exitWithError("The provided module zip file contains at least one file with an invalid name that starts with '/' or '..'. The first invalid file is named " + moduleZipMemberName + ". Please ensure that all files have valid names.")
+
+            if moduleZipMemberName.startswith(moduleName + "/"):
+                zipIncludesModuleFolder = True
+
+        if zipIncludesModuleFolder:
+            moduleZipExtrationPath = helperlib.DSC_MODULES_PATH
+        else:
+            moduleZipExtrationPath = join(helperlib.DSC_MODULES_PATH, moduleName)
+        
+        # Extract module to destination path
+        printVerboseMessage("Extracting module zip file from " + moduleZipFilePath + " to " + moduleZipExtrationPath)
+        moduleZipFile.extractall(moduleZipExtrationPath)
+    finally:
+        moduleZipFile.close()
+
+    # Check that we can find the extracted module folder
+    moduleDestinationPath = join(moduleZipExtrationPath, moduleName)
+    if not os.path.isdir(moduleDestinationPath):
+        exitWithError("Unable to find module directory after attempting to extract module from zip to the path " + moduleDestinationPath)
+
+    # Check that we can find the DSCResources folder under the extracted module folder
+    moduleDscResourcesDestinationPath = join(moduleDestinationPath, 'DSCResources')
+    if not os.path.isdir(moduleDscResourcesDestinationPath):
+        exitWithError("Unable to find the DSCResources directory under the module directory at the path " + moduleDscResourcesDestinationPath + " after extracting module from zip to the path.")
+
+    # Populate commom main DSC path
+    dscMainFolderPath = join(helperlib.CONFIG_SYSCONFDIR, helperlib.CONFIG_SYSCONFDIR_DSC) 
+
+    # Verify the module checksum if specified
+    if verifyChecksum:
+        # Check that we can find the keyring file
+        keyringFilePath = join(dscMainFolderPath, 'keyring.gpg')
+        if not os.path.isfile(keyringFilePath):
+            exitWithError("Unable to find DSC keyring file at " + keyringFilePath)
```

Comment:
the nx module zip file here /opt/microsoft/dsc/module_packages/ do not have keyring file .. are we not signing modules those are part of our package ?

Reply:
I'm not sure what the current signing process is. We do not validate modules that are built in to the package and installed through the DSC installer. Only pulled modules are currently validated. I did not make any changes to this functionality. This is the same way it functioned before